q2_dada2/_denoise.py

```diff
@@ -6,28 +6,93 @@
 
-import os.path
+import os
 import tempfile
 import hashlib
 
 import biom
 import skbio
 from q2_types.feature_data import DNAIterator
 from q2_types.per_sample_sequences import (
-    SingleLanePerSampleSingleEndFastqDirFmt)
+    FastqGzFormat, SingleLanePerSampleSingleEndFastqDirFmt,
+    SingleLanePerSamplePairedEndFastqDirFmt)
 
 from q2_dada2._plot import run_commands
 
 
-def denoise(demultiplexed_seqs: SingleLanePerSampleSingleEndFastqDirFmt,
-            trunc_len: int, trim_left: int, max_ee: int=2, truncq: int=2,
-            hashed_feature_ids: bool=True) -> (biom.Table, DNAIterator):
+def _check_featureless_table(fp):
+    with open(fp) as fh:
+        # There is a comment line and a header before the feature data
+        for line_count, _ in zip(range(3), fh):
+            pass
+    if line_count < 2:
+        raise ValueError("No features remain after denoising. "
+                         "Try adjusting your arguments.")
+
+
+def denoise_single(demultiplexed_seqs: SingleLanePerSampleSingleEndFastqDirFmt,
+                   trunc_len: int, trim_left: int=0, max_ee: float=2.0,
+                   trunc_q: int=2, n_threads: int=1,
+                   n_reads_learn: int=1000000, hashed_feature_ids: bool=True
+                   ) -> (biom.Table, DNAIterator):
     with tempfile.TemporaryDirectory() as temp_dir_name:
         biom_fp = os.path.join(temp_dir_name, 'output.tsv.biom')
-        cmd = ['run_dada.R', str(demultiplexed_seqs), biom_fp,
-               str(trunc_len), str(trim_left), str(max_ee), str(truncq),
-               temp_dir_name]
+        cmd = ['run_dada_faster.R',
+               str(demultiplexed_seqs), biom_fp, temp_dir_name,
+               str(trunc_len), str(trim_left), str(max_ee), str(trunc_q),
+               str(n_threads), str(n_reads_learn)]
+        run_commands([cmd])
+        _check_featureless_table(biom_fp)
+        table = biom.Table.from_tsv(open(biom_fp), None, None, None)
+        # Currently the sample IDs in DADA2 are the file names. We make
+        # them the sample id part of the filename here.
+        sid_map = {id_: id_.split('_')[0] for id_ in table.ids(axis='sample')}
+        table.update_ids(sid_map, axis='sample', inplace=True)
+        # The feature IDs in DADA2 are the sequences themselves.
+        if hashed_feature_ids:
+            # Make feature IDs the md5 sums of the sequences.
+            fid_map = {id_: hashlib.md5(id_.encode('utf-8')).hexdigest()
+                       for id_ in table.ids(axis='observation')}
+            table.update_ids(fid_map, axis='observation', inplace=True)
+
+            rep_sequences = DNAIterator((skbio.DNA(k, metadata={'id': v})
+                                         for k, v in fid_map.items()))
+        else:
+            rep_sequences = DNAIterator(
+                (skbio.DNA(id_, metadata={'id': id_})
+                 for id_ in table.ids(axis='observation')))
+    return table, rep_sequences
```

Comment:
Can lines 44 - 64 be moved to a single function that can be called from `denoise_single` and `denoise_paired`?